scripts/benchmark/trainer-benchmark.py

```diff
@@ -0,0 +1,435 @@
+            results_str += f" {tuple(round(x, 2) for x in results)}"
+        print(results_str)
+        mean_metrics[variation_key] = variation
+        return mean_metrics
+    else:
+        print(outcome)
+        return {variation_key: variation, target_metric_key: nan}
+
+
+def get_versions():
+    properties = torch.cuda.get_device_properties(torch.device("cuda"))
+    return f"""
+Datetime    : {datetime.datetime.now().strftime('%Y-%m-%d %H:%M:%S')}
+
+Software:
+transformers: {transformers.__version__}
+torch       : {torch.__version__}
+cuda        : {torch.version.cuda}
+python      : {platform.python_version()}
+
+Hardware:
+{torch.cuda.device_count()} GPUs      : {properties.name}, {properties.total_memory/2**30:0.2f}GB
+"""
+
+
+def process_results(results, target_metric_key, report_metric_keys, base_variation, output_dir):
+
+    df = pd.DataFrame(results)
+    variation_key = "variation"
+    diff_key = "diff_%"
+
+    sentinel_value = nan
+    if base_variation is not None and len(df[df[variation_key] == base_variation]):
+        # this may still return nan
+        sentinel_value = df.loc[df[variation_key] == base_variation][target_metric_key].item()
+    if math.isnan(sentinel_value):
+        # as a fallback, use the minimal value as the sentinel
+        sentinel_value = df.loc[df[target_metric_key] != nan][target_metric_key].min()
+
+    # create diff column if possible
+    if not math.isnan(sentinel_value):
+        df[diff_key] = df.apply(
+            lambda r: round(100 * (r[target_metric_key] - sentinel_value) / sentinel_value)
+            if not math.isnan(r[target_metric_key])
+            else 0,
+            axis="columns",
+        )
+
+    # re-order columns
+    cols = [variation_key, target_metric_key, diff_key, *report_metric_keys]
+    df = df.reindex(cols, axis="columns")  # reorder cols
+
+    # capitalize
+    df = df.rename(str.capitalize, axis="columns")
+
+    # make the cols as narrow as possible
+    df_github = df.rename(lambda c: c.replace("_", "<br>"), axis="columns")
+    df_console = df.rename(lambda c: c.replace("_", "\n"), axis="columns")
+
+    report = ["", "Copy between the cut-here-lines and paste as is to github or a forum"]
+    report += ["----------8<-----------------8<--------"]
+    report += ["*** Results:", df_github.to_markdown(index=False, floatfmt=".2f")]
+    report += ["```"]
+    report += ["*** Setup:", get_versions()]
+    report += ["*** The benchmark command line was:", get_orig_cmd()]
+    report += ["```"]
+    report += ["----------8<-----------------8<--------"]
+    report += ["*** Results (console):", df_console.to_markdown(index=False, floatfmt=".2f")]
+
+    print("\n\n".join(report))
+
+
+def main():
+    parser = argparse.ArgumentParser()
+    parser.add_argument(
+        "--base-cmd",
+        default=None,
+        type=str,
+        required=True,
+        help="Base cmd",
+    )
+    parser.add_argument(
+        "--variations",
```

Comment:
do the variations just correspond to the precision? Maybe call it `precision_variations` then?

Reply:
no. 

Variations are just that - variations - it can be anything. `--variations "A|B|C" "D|E|F"` will do a Cartesian product to produce `A D`, `A E`, ..., `C F` and run each variation.

I was just using precisions as an example. but you can test say  `--variations "--bs 4|--bs 6|--bs 8"` (we don't have `--bs`, but you get the idea).

I'm open to suggestions if the name is not intuitive.

Reply:
I see! Sorry I misread it then. But then variations can only correspond to args of `TrainingArguments` no? So maybe we can say `training_args_variations`?

Reply:
The idea is that this tool is going to be expanded to support other programs which don't necessarily have training args, after all it doesn't care which args are varied. @siddk is planning to expand it to support `accelerate` and down the road others, hence the generic name.

So the mnemonic is "variations to compare" whatever they are.